execution_engine/src/core/runtime/mint_internal.rs

```diff
@@ -189,4 +189,31 @@ where
     R: StateReader<Key, StoredValue>,
     R::Error: Into<execution::Error>,
 {
+    /// Burns native tokens.
+    fn burn(&mut self, purse: URef, amount: U512) -> Result<(), Error> {
+        let purse_key = Key::URef(purse);
+        self.context
+            .validate_writeable(&purse_key)
+            .map_err(|_| Error::InvalidAccessRights)?;
+        self.context
+            .validate_key(&purse_key)
+            .map_err(|_| Error::InvalidURef)?;
+
+        let source_balance: U512 = match self.read_balance(purse)? {
+            Some(source_balance) => source_balance,
+            None => return Err(Error::PurseNotFound),
+        };
+
+        let new_balance = match source_balance.checked_sub(amount) {
+            Some(value) => value,
+            None => U512::zero(),
```

Comment:
I think we may want to be explicit here and disallow burning more than the purse balance has - instead, we can return an error. My reasoning here is we already don't allow transferring more funds than a source purse has (source: https://github.com/casper-network/casper-node/blob/5b55aa1c7c990cabc43bb264f4c44644afaf7d1e/execution_engine/src/system/mint.rs#L220-L226). Then later, you shouldn't need to calculate `burned_amount` as you can reduce total supply by `amount` (cc @EdHastingsCasperAssociation)

```suggestion
            None => return Err(Error::InsufficientFunds),
```

Since this method will return a `Result<(), mint::Error>` make sure `call_host_mint` returns a `Result` too (and this will be consistent with other mint's entrypoints)

Reply:
i would think an attempt to burn more than the balance would just burn the remaining balance instead. in other words if a purse had balance 30 and attempt to burn 31 would burn 30, zeroing the purse.